packages/jupyter-ai/src/components/chat.tsx

```diff
@@ -143,6 +146,21 @@ function ChatBody({
     );
   }
 
+  const myHumanMessageIds = new Set(
+    messages
+      .filter(
+        m => m.type === 'human' && m.client.username === user?.identity.username
+      )
+      .map(m => m.id)
+  );
+
+  const myStreamedMessages = messages.filter(
+    m =>
+      m.type === 'agent-stream' &&
+      !m.complete &&
+      myHumanMessageIds.has(m.reply_to)
+  ) as AiService.AgentStreamMessage[];
+
```

Comment:
> we don't need to maintain a list of all the replies that are still streaming.

I don't follow, we do maintain a list of all messages either way in order to accept streams, right? So why not?

> My concern with the addition of this logic is that the list of messages is being filtered twice on every render of the `<Chat />` component (which occurs on the receipt of every stream chunk). This can have a significant performance drawback.

I don't believe this would be a significant performance drawback. Even if there was 1 million messages it would take less than 10 ms to filter it on a laptop (I just tried it, [here](https://jsbenchmark.com/#eyJjYXNlcyI6W3siaWQiOiJ6d1l1bUJhSTZPdTBadllVWFFPYl8iLCJjb2RlIjoiY29uc3QgbWVzc2FnZXMgPSBEQVRBO1xuY29uc3QgbXlIdW1hbk1lc3NhZ2VJZHMgPSBuZXcgU2V0KFxuICBtZXNzYWdlc1xuICAgIC5maWx0ZXIoXG4gICAgICBtID0-IG0udHlwZSA9PT0gJ2h1bWFuJyAmJiBtLmNsaWVudC51c2VybmFtZSA9PT0gdXNlcj8uaWRlbnRpdHkudXNlcm5hbWVcbiAgICApXG4gICAgLm1hcChtID0-IG0uaWQpXG4pO1xuXG5jb25zdCBteVN0cmVhbWVkTWVzc2FnZXMgPSBtZXNzYWdlcy5maWx0ZXIoXG4gIG0gPT5cbiAgICBtLnR5cGUgPT09ICdhZ2VudC1zdHJlYW0nICYmXG4gICAgIW0uY29tcGxldGUgJiZcbiAgICBteUh1bWFuTWVzc2FnZUlkcy5oYXMobS5yZXBseV90bylcbik7IiwibmFtZSI6IkZpbmQgOTkiLCJkZXBlbmRlbmNpZXMiOltdfV0sImNvbmZpZyI6eyJuYW1lIjoiQmFzaWMgZXhhbXBsZSIsInBhcmFsbGVsIjp0cnVlLCJnbG9iYWxUZXN0Q29uZmlnIjp7ImRlcGVuZGVuY2llcyI6W119LCJkYXRhQ29kZSI6InJldHVybiBBcnJheSgxMDAwMDAwKS5tYXAoKGkpID0-IChcbiAgTWF0aC5yYW5kb20oKSA-IDAuNVxuICA_IHtcbiAgICB0eXBlOiAnaHVtYW4nLFxuICAgIGNsaWVudDoge3VzZXJuYW1lOiBNYXRoLnJhbmRvbSgpID4gMC41ID8gJ215c2VsZicgOiAnY29sbGFib3JhdG9yJ30sXG4gICAgaWQ6IGlcbiAgfVxuICA6IHtcbiAgICB0eXBlOiAnYWdlbnQtc3RyZWFtJyxcbiAgICBjb21wbGV0ZTogTWF0aC5yYW5kb20oKSA-IDAuNSxcbiAgICByZXBseV90bzogaSAtIDFcbiAgfVxuKSkifX0)).

Reply:
> @krassowski Can you give me contributor permissions on your fork so I can open a PR against this branch that includes the changes I'm proposing?

You can either push to this branch directly or open a PR against my fork already :)

Reply:
I've pushed a couple of commits to your branch. I'll leave the details in a separate comment.